### PR TITLE
Correct 2-D FE_Nedelec Interpolation

### LINKED ISSUE #328: Incorrect behaviour of FE_Nedelec::interpolate
_From [agray...@gfz-potsdam.de](https://code.google.com/u/116043767735290287173/) on July 05, 2014 00:20:34_

It seems that FE_Nedelec::interpolate does not work correctly for high orders. I have replicated the interpolate_rt.cc test in the attached file  and the output it gives is:

FE_Nedelec<3>(0) 12  vector 1.66533e-16 Vector 1.66533e-16
FE_Nedelec<3>(1) 56  vector 2.77556e-17 Vector 2.77556e-17
FE_Nedelec<3>(2) 117  vector 0.128996 Vector 0.177798
FE_Nedelec<3>(3) 208  vector 0.241009 Vector 0.32687

As far as I understand this indicates that for high-order elements something goes wrong, does not it? 
Any idea what could that be?

Thank you,
Alexander

**Attachment:** [interpolate_nedelec.cc](http://code.google.com/p/dealii/issues/detail?id=231)

_Original issue: http://code.google.com/p/dealii/issues/detail?id=231_

### PR DESCRIPTION
**Pull request by harmonj**
Fixes a conditional statement in FE_Nedelec::convert_...to_dof_values. Adds a new test function in tests/fe/interpolation_common.h for use in tests/fe/interpolation_nedelec.cc as the previous test case indicated (incorrectly) that the function returned correct DoF values.

~Does not fix~ Related to #328, which concerns the 3-D element. (edit by @drwells: don't autoclose the cited issue)

**Output before (with updated test case):**
```
DEAL::FE_Nedelec<2>(0) 4  vector 2.77556e-17
DEAL::FE_Nedelec<2>(1) 12  vector 1.38778e-17
DEAL::FE_Nedelec<2>(2) 21  vector 5.55112e-17
DEAL::FE_Nedelec<2>(2) 21  vector 1.11022e-16
DEAL::FE_Nedelec<2>(2) 21  vector 2.81025e-16
DEAL::FE_Nedelec<2>(3) 32  vector 7.56830e-16
DEAL::FE_Nedelec<2>(1) 12  vector 0.387298
DEAL::FE_Nedelec<2>(2) 21  vector 4.44089e-16
DEAL::FE_Nedelec<2>(3) 32  vector 1.60288e-15
DEAL::FE_Nedelec<3>(1) 56  vector 2.77556e-17
DEAL::FE_Nedelec<3>(2) 117  vector 1.66533e-16
```

**Updated output:**
```
DEAL::FE_Nedelec<2>(0) 4  vector 2.77556e-17
DEAL::FE_Nedelec<2>(1) 12  vector 1.38778e-17
DEAL::FE_Nedelec<2>(2) 21  vector 5.55112e-17
DEAL::FE_Nedelec<2>(2) 21  vector 1.11022e-16
DEAL::FE_Nedelec<2>(2) 21  vector 2.81025e-16
DEAL::FE_Nedelec<2>(3) 32  vector 7.56830e-16
DEAL::FE_Nedelec<2>(1) 12  vector 5.55112e-17
DEAL::FE_Nedelec<2>(2) 21  vector 4.44089e-16
DEAL::FE_Nedelec<2>(3) 32  vector 1.60288e-15
DEAL::FE_Nedelec<3>(1) 56  vector 2.77556e-17
DEAL::FE_Nedelec<3>(2) 117  vector 1.66533e-16
```

### CODE REVIEW

File: source/fe/fe_nedelec.cc [hunk 1]
```diff
@@ -3201,7 +3201,7 @@ FE_Nedelec<dim>::convert_generalized_support_point_values_to_dof_values(
           // are obtained by solving
           // a linear system of
           // equations.
-          if (this->degree - 1 > 1)
+          if (this->degree > 1)
```
Comment (drwells) [inline source/fe/fe_nedelec.cc:3204]: Does this fix imply that the comment above is also wrong?

**Comment by bangerth** [comment]
@harmonj Thanks for starting to upstream your patches -- I had always meant to come back to asking you about that, but it never percolated to the top of my todo list :-( We look forward to more patches!

**Comment by bangerth** [comment]
Separately from @drwells 's comment above, can you also run `./contrib/utilities/indent` in your source directory and follow the instructions it gives you if there is an error? One of the automated checks complains about an indentation problem in your test.

File: source/fe/fe_nedelec.cc (hunk 1 above)
Comment (harmonj) [reply source/fe/fe_nedelec.cc:3204]: The comment above, though confusing, is technically correct. The comment refers to the degree of the FE_Nedelec cell, whereas this->degree returns the highest polynomial degree of the cell (which is FE_Nedelec_degree + 1).

**Comment by harmonj** [comment]
> Separately from @drwells 's comment above, can you also run `./contrib/utilities/indent` in your source directory and follow the instructions it gives you if there is an error? One of the automated checks complains about an indentation problem in your test.

Yes, I ran it before, but after I deleted a comment I must've ruined the indentation.

I ran it again and pushed the change.

File: source/fe/fe_nedelec.cc (hunk 1 above)
Comment (bangerth) [reply source/fe/fe_nedelec.cc:3204]: Ah, interesting. Would you mind clarifying the comment in this regard by pointing out the difference between the degree of the element and that `this->degree` is the *polynomial* degree?

I suspect that it is exactly this difference that led to the bug in the first place.

**Comment by bangerth** [comment]
One last step: When you have all changes made, could you squash the commits into one?

**Comment by bangerth** [comment]
Oh, actually, not all done: You should place a file into `doc/news/changes/minor` that describes the fix! Just take any of the existing files in that directory as an example. This way you will be listed as one of the authors of deal.II!

**Comment by harmonj** [comment]
I made all changes above and squashed into one commit.

Thank you, @bangerth and @drwells for the comments/help.

**Review by drwells — APPROVED** [review]

**Review by bangerth — APPROVED** [review]
Excellent, thank you! I'm looking forward to your other patches!

**Pull request merged**